062-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,202 @@
+- `YammerPrometheusMetricsReporter` usable on brokers (for Yammer metrics)
+
+The Prometheus metrics registry is a singleton and the HTTP server will also be a singleton. This will allow applications to start multiple instances of the reporter (for example in applications containing multiple Kafka clients like Streams, Connect), and still collect all metrics via a single HTTP endpoint per JVM.
+
+The reporter for Kafka metrics will be usable outside of Strimzi by applications using Kafka clients. To do so applications will need to set the `metric.reporters` configuration to `KafkaPrometheusMetricsReporter` and set the reporter configurations accordingly for each Kafka client they instantiate.
+
+### Handling of non-numeric metrics
+
+Out of all the metrics emitted by Kafka brokers and clients, a few of them have non-numeric values. Prometheus only supports numeric values for metrics. When using jmx_exporter it is possible to write rules that move the non-numeric value to a label. For example we do it in [kafka-connect-metrics](https://github.com/strimzi/strimzi-kafka-operator/blob/main/packaging/examples/metrics/kafka-connect-metrics.yaml#L103-L111) for the `status` metrics which has a string value.
+
+I propose to do this automatically in the reporters. For example a metric named: `kafka.connect:type=connector-task-metrics,connector="{connector}",task="{task}"<>status` with the value `running` will be converted into `kafka_connect_connector_task_metrics_status{connector="heartbeats",task="0",status="running"}` and its value will be set to `1.0`.
+
+## Affected/not affected projects
+
+The reporter will be a new project, `strimzi-metrics-reporter`. It will be used by Strimzi components but should also be usable without Strimzi, for example in client side applications.
+
+### strimzi-kafka-operator
+
+Reporters are not usable with ZooKeeper so Strimzi should keep support for jmx_exporter while it also supports ZooKeeper. As it's new code, reporters should first be released as a preview feature and disabled by default. Once KRaft becomes the preferred deployment option and the, we should make the reporter the recommended and default option to collect metrics over jmx_exporter.
+
+We will need new examples demonstrating how to use the reporters, new grafana dashboards with metric names from the reporters.
+
+### strimzi-kafka-oauth
+
+The OAuth plugin exposes the `strimzi.oauth.metric.reporters` configuration to let users specify a metrics reporter. Today the OAuth plugin automatically uses `org.apache.kafka.common.metrics.JmxReporter` if that configuration is not set. To switch to the new reporter, users should set `strimzi.oauth.metric.reporters` to `KafkaPrometheusMetricsReporter`. Once the reporter becomes the preferred option, the OAuth plugin should be updated to automatically use `KafkaPrometheusMetricsReporter` if no metrics reporters are set by users.
```

Comment:
The OAuth plugin can be used independently on Strimzi. So I'm not sure if defaulting its options to some other Strimzi dependency is the right way. It should IMHO stick with what Kafka has and if needed, we can configure the value of this option through the operator and the custom resources in Strimzi-based deployments.

Reply:
You are right, I'll remove that last sentence as it makes no sense when the plugin is used outside of Strimzi.

Reply:
Done, resolving this comment